File: USACO_training/sprime.cpp
/*
ID: s.sophi1
LANG: C++14 
PROG: sprime             
*/

#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <utility> 
#include <math.h>

using namespace std;

ofstream fout ("sprime.out");

bool is_prime(int x) {
	for (int i = 2; i <= sqrt(x); i++) {
		if (x % i == 0) return false;
	}
	return true;
}

vector<int> results;

int is_super_prime(int number) {
	int num_len = to_string(number).length();
	for (int i = 0; i < num_len; i++) {
		if (is_prime(number)) 
			number = number/10;
		else return false;
	}
	return true;
}

int potential_super_primes(int length, string curr_num) {
	if (curr_num.length() == length) {
		if (is_super_prime(atoi(curr_num.c_str())))
			fout << curr_num << endl;
	}
	else {
		string first_num[4] = {"2", "3", "5", "7"};
		string not_first_num[4] = {"1", "3", "7", "9"};
		if (curr_num.length() == 0) {
			for (const string num: first_num) {
				potential_super_primes(length, curr_num+num);
			}
		}
		else {
			for (const string num: not_first_num) {
				potential_super_primes(length, curr_num+num);
			}
		}
	}
}
int main() {
	ifstream fin ("sprime.in");
	int length;
	fin >> length;
	potential_super_primes(length, "");

	return 0;
}

File: USACO_training/sprime.py
"""
ID: s.sophi1
LANG: PYTHON3
PROG: sprime
"""

from math import sqrt
input = open('sprime.in', 'r')
output = open('sprime.out', 'w')
strings = input.readlines()
strings = [string.rstrip() for string in strings]
input.close()


def is_prime(number):
    if number == 1 or number == 0:
        return False
    for i in range(2, int(sqrt(number)+1)):
        if number % i == 0:
            return False
    return True


def is_super_prime(number):
    num_len = len(str(number))

    for i in range(num_len):
        if is_prime(number):
            number = number//10
        else:
            return False
    return True


def potential_super_primes(length, curr_num):
    if len(curr_num) == length:
        if is_super_prime(int(curr_num)):
            output.write(curr_num+'\n')
    else:
        first_num = ['2','3','5','7']
        not_first_num = ['1','3','7','9']
        if len(curr_num) == 0:
            for num in first_num:
                potential_super_primes(length, curr_num+num)
        else:
            for num in not_first_num:
                potential_super_primes(length, curr_num+num)


potential_super_primes(length, '')
output.close()
